### DISCUSSION TITLE
Need to check if we can create device hierarchy in edgexfoundry

### DISCUSSION THREAD

**Discussion by PushpamTiwari** [edgexfoundry/community#178] [Q&A]
I am new in edgex foundry. I have one question Can I create a device hierarchy in Edgex Foundry?

Example: I want to create a device hierarchy for the Rail system with 10 trains.

Root Device: RailTail
  Child Device: Hyd Metro
    Child Device: Tracks
      Child Device: Train[1-10]
        Child Device: DC Train
          Child Device: Bogie
            Child Device: Weather Sensor
              Property: Temp
              Property: Humidity
            Child Device: Doors
              Property: Open/Close
              Property: Count (isOpen / isClose)
              Property: Count no of time door open/close
              Property: Standing before the door
          Child Device: Engine
            Child Device: Weather Sensor
              Property: Temp
              Property: Humidity
            Child Device: Pressure/RPM
            
  **OR SECOND APPROACH** 
  
  
  Train 1
├── DC Train
│   ├── Bogie
│   │   ├── Weather Sensor (Temperature, Humidity)
│   │   ├── Doors
│   │   │   ├── Close/Open Sensor (isOpen/isClose)
│   │   │   ├── Door Counter
│   │   │   ├── Passenger Detector
│   │   └── ...
│   ├── Engine
│   │   ├── Weather Sensor (Temperature, Humidity)
│   │   ├── Pressure Sensor
│   │   ├── RPM Sensor
│   │   └── ...
└── ...

Train 2
├── DC Train
│   ├── Bogie
│   │   ├── ...
│   ├── Engine
│   │   ├── ...
└── ...

(Repeat for 10 trains)

**Comment by cloudxxx8** [comment]
Is there any hierachy from the device protocol point of view?

**Reply by PushpamTiwari** [reply]
no there is no hierarchy for device protocol, I mean I am still figuring it out. for inter-communication, the protocol will be OPCA or MQTT.

**Reply by cloudxxx8** [reply]
There is a new `properties` field which can store any additional information you need in v3.
https://github.com/edgexfoundry/go-mod-core-contracts/blob/a3b42e94f94e881b46b1ffd7dc1266b644a738ca/dtos/device.go#L26
You can store the parent device name in the properties of each child device to build the hierachy.
